server/src/main/java/org/elasticsearch/index/mapper/PostParsePhase.java

```diff
@@ -0,0 +1,498 @@
+        public NumericDocValues getNumericDocValues(String field) throws IOException {
+            checkField(field);
+            List<Number> values = document.getFields().stream()
+                .filter(f -> Objects.equals(f.name(), field))
+                .filter(f -> f.fieldType().docValuesType() == DocValuesType.NUMERIC)
+                .map(IndexableField::numericValue)
+                .sorted()
+                .collect(Collectors.toList());
+            return numericDocValues(values);
+        }
+
+        @Override
+        public BinaryDocValues getBinaryDocValues(String field) throws IOException {
+            checkField(field);
+            List<BytesRef> values = document.getFields().stream()
+                .filter(f -> Objects.equals(f.name(), field))
+                .filter(f -> f.fieldType().docValuesType() == DocValuesType.BINARY)
+                .map(IndexableField::binaryValue)
+                .sorted()
+                .collect(Collectors.toList());
+            return binaryDocValues(values);
+        }
+
+        @Override
+        public SortedDocValues getSortedDocValues(String field) throws IOException {
+            checkField(field);
+            List<BytesRef> values = document.getFields().stream()
+                .filter(f -> Objects.equals(f.name(), field))
+                .filter(f -> f.fieldType().docValuesType() == DocValuesType.SORTED)
+                .map(IndexableField::binaryValue)
+                .sorted()
+                .collect(Collectors.toList());
+            return sortedDocValues(values);
+        }
+
+        @Override
+        public SortedNumericDocValues getSortedNumericDocValues(String field) throws IOException {
+            checkField(field);
+            List<Number> values = document.getFields().stream()
+                .filter(f -> Objects.equals(f.name(), field))
+                .filter(f -> f.fieldType().docValuesType() == DocValuesType.SORTED_NUMERIC)
+                .map(IndexableField::numericValue)
+                .sorted()
+                .collect(Collectors.toList());
+            return sortedNumericDocValues(values);
+        }
+
+        @Override
+        public SortedSetDocValues getSortedSetDocValues(String field) throws IOException {
+            List<BytesRef> values = document.getFields().stream()
+                .filter(f -> Objects.equals(f.name(), field))
+                .filter(f -> f.fieldType().docValuesType() == DocValuesType.SORTED_SET)
+                .map(IndexableField::binaryValue)
+                .sorted()
+                .collect(Collectors.toList());
+            return sortedSetDocValues(values);
+        }
+
+        @Override
+        public FieldInfos getFieldInfos() {
+            return new FieldInfos(new FieldInfo[0]);
+        }
+
+        @Override
+        public void document(int docID, StoredFieldVisitor visitor) throws IOException {
+            visitor.binaryField(SOURCE_FIELD_INFO, sourceBytes.toBytesRef().bytes);
```

Comment:
Do we need to support other stored fields here, since we support `_fields` in runtime field scripts?

Reply:
We do! Will add that in.

Reply:
So this is possible, but it is not trivial as we need to do some careful handling of different types.  I'm now wondering if we should go the other way and remove stored fields access from runtime scripts - I know @jimczi was thinking that we should be discouraging stored fields use in general.

Reply:
I have implemented this properly with a test in MapperTestCase to ensure that it works for all mappers - turns out it wasn't quite as tricky as I thought, although it did involve a minor change to LeafStoredFieldsLookup to improve typing.